File: JasSwift/Sources/Panels/ConceptsPanel.swift
import Foundation

/// Native action arm for the Concepts panel (CONCEPTS.md §6). The
/// `concepts_panel_select` action is generic (`set_panel_state` writes
/// `selected_concept` into the shared store); `place_concept_instance` is a
/// `log` stub whose real work — mint a fresh element id (value-in-op), build a
/// Generated element from the panel-selected concept + its declared default
/// params, append + select — lives here, mirroring SymbolsPanel and the Rust
/// dispatch arm.
public enum ConceptsPanel {
    /// The StateStore scope id for the Concepts panel; the generic
    /// `set_panel_state { panel: concepts }` effect appends `_panel_content`,
    /// and the YAML `panel.selected_concept` read resolves against it.
    private static let scopeId = "concepts_panel_content"

    /// The panel-selected concept id from the shared store, or nil.
    static func selectedConcept(_ model: Model) -> String? {
        model.stateStore.getPanel(scopeId, "selected_concept") as? String
    }

    /// The concept's declared default params as a params object.
    static func defaultParams(_ conceptId: String) -> [String: Any] {
        guard let concept = WorkspaceData.load()?.concept(conceptId),
              let params = concept["params"] as? [[String: Any]] else { return [:] }
        var out: [String: Any] = [:]
        for p in params {
            if let name = p["name"] as? String, let def = p["default"] {
                out[name] = def
            }
        }
        return out
    }

    /// Gather every existing element id (layers + master store) so a freshly
    /// minted id avoids collisions. Mirrors SymbolsPanel.existingIds.
    private static func existingIds(_ doc: Document) -> Set<String> {
        var set: Set<String> = []
        func gather(_ elem: Element) {
            if let id = elem.id { set.insert(id) }
            switch elem {
            case .group(let g): for c in g.children { gather(c) }
            case .layer(let l): for c in l.children { gather(c) }
            default: break
            }
        }
        for layer in doc.layers { gather(.layer(layer)) }
        for master in doc.symbols { gather(master) }
        return set
    }

    private static func mint(_ existing: Set<String>) -> String? {
        for _ in 0..<100 {
            let c = generateElementId()
            if !existing.contains(c) { return c }
        }
        return nil
    }

    /// Native intercept for the Concepts panel ops (the YAML actions are
    /// `log` stubs). Routes through `opApply` so the placement JOURNALS as a
    /// real `place_concept_instance` op (value-in-op: the panel-selected
    /// concept id + its RESOLVED default params + the minted elem id),
    /// replayable like the sibling structural verbs. `withTxn` brackets the one
    /// undo step; the arm both mutates and records. Mirrors the Rust dispatch
    /// arm and the native-menu delete routing.
    public static func dispatch(_ action: String, model: Model) {
        switch action {
        case "place_concept_instance":
            guard let conceptId = selectedConcept(model) else { return }
            let existing = existingIds(model.document)
            guard let elemId = mint(existing) else { return }
            let op: [String: Any] = [
                "op": "place_concept_instance",
                "concept_id": conceptId,
                "params": defaultParams(conceptId),
                "elem_id": elemId,
            ]
            model.withTxn {
                model.nameTxn("place_concept_instance")
                opApply(model, Controller(model: model), op)
            }
        case "promote_to_concept":
            // CONCEPTS.md §10 — the fitter / promote (the inverse of expand).
            // Detect + replace the single selected raw shape with a Generated.
            promote(model: model)
        default:
            break
        }
    }

    /// Native intercept for `set_concept_param` (the YAML action is a `log`
    /// stub): find the single selected Generated instance and write
    /// `params[name] = value`. Routes through `opApply` so the edit JOURNALS as
    /// a real `set_concept_param` op (value-in-op: the RESOLVED path, param
    /// name, and committed value), replayable like the sibling property verbs.
    /// `withTxn` brackets the one undo step. Mirrors the Rust `set_concept_param`
    /// dispatch arm. No-op unless exactly one Generated element is selected.
    public static func setParam(model: Model, name: String, value: Double) {
        let doc = model.document
        guard doc.selection.count == 1, let sel = doc.selection.first else { return }
        let path = sel.path
        guard case .live(.generated) = doc.tryGetElement(path) else { return }
        let op: [String: Any] = [
            "op": "set_concept_param",
            "path": path,
            "name": name,
            "value": value,
        ]
        model.withTxn {
            model.nameTxn("set_concept_param")
            opApply(model, Controller(model: model), op)
        }
    }

    /// Native intercept for `apply_concept_operation` (CONCEPTS.md §9): apply a
    /// named concept operation to the single selected Generated instance. The
    /// operation's effect is RESOLVED here, at production time — look the
    /// operation up in the registry by `opId`, evaluate its `set:` expressions
    /// with the instance's CURRENT params bound under `param`, and bake the
    /// resulting `changes` map into the op (value-in-op). Routed through
    /// `opApply` inside the one-undo `withTxn`/`nameTxn` bracket; replay merges
    /// `changes` and never re-evaluates. Mirrors the Rust `apply_concept_operation`
    /// dispatch arm. No-op unless exactly one Generated element is selected and
    /// the resolved changes are non-empty.
    public static func applyOperation(model: Model, opId: String) {
        let doc = model.document
        guard doc.selection.count == 1, let sel = doc.selection.first else { return }
        let path = sel.path
        guard case .live(.generated(let gen)) = doc.tryGetElement(path) else { return }

        // Look up the concept + operation, then resolve its `set:` expressions
        // over the instance's current params → the concrete `changes` map.
        guard let concept = WorkspaceData.load()?.concept(gen.conceptId),
              let ops = concept["operations"] as? [[String: Any]],
              let operation = ops.first(where: { ($0["id"] as? String) == opId }),
              let set = operation["set"] as? [String: Any] else { return }

        let ctx: [String: Any] = ["param": gen.params]
        var changes: [String: Any] = [:]
        for (name, exprV) in set {
            guard let src = exprV as? String else { continue }
            if case .number(let n) = evaluate(src, context: ctx) {
                changes[name] = n
            }
        }
        guard !changes.isEmpty else { return }

        let op: [String: Any] = [
            "op": "apply_concept_operation",
            "path": path,
            "op_id": opId,
            "changes": changes,
        ]
        model.withTxn {
            model.nameTxn("apply_concept_operation")
            opApply(model, Controller(model: model), op)
        }
    }

    /// Native intercept for `promote_to_concept` (CONCEPTS.md §10 — the fitter /
    /// promote, the inverse of expand). Promote the single selected raw shape to
    /// a `Generated` concept instance: extract the element's WORLD-space vertices
    /// (bake any element transform into the points so the fitter sees world
    /// space), try each registered concept's `fitter` expression over them (bound
    /// under `shape.points`) in sorted-id order, and on the FIRST match split its
    /// flat result `[params..., cx, cy, rotation]` into the concept params (first
    /// K, by declared order) and a placement transform (`translate(cx,cy) ·
    /// rotate(rot)`). Everything is baked into the op value-in-op and routed
    /// through `opApply` in the one-undo `withTxn`/`nameTxn` bracket; a no-match
    /// (or a non-Polygon/Polyline selection) is a silent no-op. Mirrors the Rust
    /// `promote_to_concept` dispatch arm.
    public static func promote(model: Model) {
        let doc = model.document
        guard doc.selection.count == 1, let sel = doc.selection.first else { return }
        let path = sel.path
        guard let elem = doc.tryGetElement(path) else { return }

        // Only a Polygon / Polyline carries promotable vertices in v1.
        let rawPoints: [(Double, Double)]
        switch elem {
        case .polygon(let p): rawPoints = p.points
        case .polyline(let p): rawPoints = p.points
        default: return
        }
        // Bake any element transform into the points so the fitter sees WORLD
        // space (the promoted instance re-places via its own transform).
        let pts: [(Double, Double)]
        if let t = elem.transform {
            pts = rawPoints.map { t.applyPoint($0.0, $0.1) }
        } else {
            pts = rawPoints
        }
        let pointsList: [[Double]] = pts.map { [$0.0, $0.1] }
        let ctx: [String: Any] = ["shape": ["points": pointsList]]

        // Try each registered concept's fitter in sorted-id order (a
        // deterministic first-match); keep the first that matches.
        guard let registry = WorkspaceData.load()?.concepts() else { return }
        var chosen: (conceptId: String, params: [String: Any],
                     cx: Double, cy: Double, rot: Double)?
        for id in registry.keys.sorted() {
            guard let concept = registry[id] as? [String: Any],
                  let fitter = concept["fitter"] as? String else { continue }
            // Null / non-list ⇒ no match for this concept.
            guard case .list(let items) = evaluate(fitter, context: ctx) else { continue }
            // The concept's declared params, in order — the first K result slots.
            let paramNames: [String] = (concept["params"] as? [[String: Any]] ?? [])
                .compactMap { $0["name"] as? String }
            let k = paramNames.count
            guard items.count >= k + 3 else { continue }  // need params + cx,cy,rot
            let nums: [Double] = items.map { fitterNum($0.value) }
            var params: [String: Any] = [:]
            for (i, name) in paramNames.enumerated() { params[name] = nums[i] }
            chosen = (id, params, nums[k], nums[k + 1], nums[k + 2])
            break
        }
        guard let pick = chosen else { return }  // nothing matched: no-op

        // Placement: translate(cx,cy) * rotate(rot) — rotate then translate.
        let t = Transform.translate(pick.cx, pick.cy)
            .multiply(Transform.rotate(pick.rot))
        let op: [String: Any] = [
            "op": "promote_to_concept",
            "path": path,
            "concept_id": pick.conceptId,
            "params": pick.params,
            "transform": [t.a, t.b, t.c, t.d, t.e, t.f],
        ]
        model.withTxn {
            model.nameTxn("promote_to_concept")
            opApply(model, Controller(model: model), op)
        }
    }
}

/// Read a Double out of an expression-list item (NSNumber / Double / Int),
/// defaulting to 0 (the non-crashing form), for the fitter result splitting.
private func fitterNum(_ v: Any) -> Double {
    if let d = v as? Double { return d }
    if let n = v as? NSNumber { return n.doubleValue }
    if let i = v as? Int { return Double(i) }
    return 0
}

File: jas_ocaml/lib/panels/concepts_panel.ml
(** Concepts panel native glue (CONCEPTS.md section 6). The panel body (concept
    row list + Place footer) is rendered by the generic YAML interpreter from
    [workspace/panels/concepts.yaml] — a [foreach] over [data.concepts]. This
    module supplies the native Place arm (which BUILDS the VALUE-IN-OP
    [place_concept_instance] op for the panel view to route through
    [Op_apply.op_apply]) and the render-time concept resolver.

    Panel-selection ([selected_concept]) is a single concept id (or none),
    written into the panel's State_store scope by the GENERIC
    [concepts_panel_select] ([set_panel_state]); the place arm reads it. *)

let content_id = "concepts_panel_content"
let selected_key = "selected_concept"

(* The compiled workspace, loaded once — concepts are static registry data. *)
let workspace = lazy (Workspace_loader.load ())

(** The panel-selected concept id, or [None] when none is selected. *)
let selected_concept (store : State_store.t) : string option =
  match State_store.get_panel store content_id selected_key with
  | `String s when s <> "" -> Some s
  | _ -> None

(* Gather every existing element id so a freshly minted id avoids collisions
   (mirrors Symbols_panel.existing_ids). *)
let existing_ids (doc : Document.document) : (string, unit) Hashtbl.t =
  let set = Hashtbl.create 16 in
  let rec gather elem =
    (match Element.id_of elem with
     | Some id -> Hashtbl.replace set id ()
     | None -> ());
    match elem with
    | Element.Group { children; _ } | Element.Layer { children; _ } ->
      Array.iter gather children
    | _ -> ()
  in
  Array.iter gather doc.Document.layers;
  Array.iter gather doc.Document.symbols;
  set

(* Mint a collision-free stable id, retrying up to 100 times (mirrors
   Symbols_panel.mint / the Rust mint loop). *)
let mint (existing : (string, unit) Hashtbl.t) : string option =
  let rec loop n =
    if n <= 0 then None
    else
      let c = Element.generate_id () in
      if Hashtbl.mem existing c then loop (n - 1) else Some c
  in
  loop 100

(** The declared default parameters of [concept_id] as a JSON object
    [{ name -> default }] from the registry; [`Assoc []] when the concept or its
    params are missing. *)
let default_params (concept_id : string) : Yojson.Safe.t =
  match Lazy.force workspace with
  | None -> `Assoc []
  | Some ws ->
    (match Workspace_loader.concept ws concept_id with
     | None -> `Assoc []
     | Some spec ->
       (match Workspace_loader.json_member "params" spec with
        | Some (`List params) ->
          `Assoc (List.filter_map (fun p ->
            match Workspace_loader.json_member "name" p,
                  Workspace_loader.json_member "default" p with
            | Some (`String name), Some default -> Some (name, default)
            | _ -> None) params)
        | _ -> `Assoc []))

(** PLACE INSTANCE: build the VALUE-IN-OP [place_concept_instance] op for the
    panel-selected concept (CONCEPTS.md section 6-7) — the concept id, its
    RESOLVED default params (from the registry, baked in so replay never
    re-consults it), and a freshly minted element id. [None] when no concept is
    selected (or the id space is exhausted). The caller brackets one undo and
    routes the op through [Op_apply.op_apply] so it both mutates AND journals,
    replayable like the sibling structural verbs. *)
let place_concept_op (store : State_store.t) (m : Model.model)
  : Yojson.Safe.t option =
  match selected_concept store with
  | None -> None
  | Some concept_id ->
    let existing = existing_ids m#document in
    (match mint existing with
     | None -> None
     | Some elem_id ->
       let params = default_params concept_id in
       Some (`Assoc [
         ("op", `String "place_concept_instance");
         ("concept_id", `String concept_id);
         ("params", params);
         ("elem_id", `String elem_id);
       ]))

(** SET PARAM: build the VALUE-IN-OP [set_concept_param] op that writes [value]
    onto parameter [name] of the single selected Generated instance so it
    re-generates live (CONCEPTS.md section 6.4). [None] unless exactly one
    Generated element is selected. The path / name / value are baked into the op
    (resolved at production time, never re-derived on replay). The caller
    brackets one undo and routes through [Op_apply.op_apply]. *)
let set_concept_param_op (_store : State_store.t) (m : Model.model)
    (name : string) (value : float) : Yojson.Safe.t option =
  let doc = m#document in
  match Document.PathMap.bindings doc.Document.selection with
  | [ (path, _) ] ->
    (match (try Some (Document.get_element doc path) with _ -> None) with
     | Some (Element.Live (Element.Generated _)) ->
       Some (`Assoc [
         ("op", `String "set_concept_param");
         ("path", `List (List.map (fun i -> `Int i) path));
         ("name", `String name);
         ("value", `Float value);
       ])
     | _ -> None)
  | _ -> None

(** APPLY OPERATION: build the VALUE-IN-OP [apply_concept_operation] op for the
    named operation [op_id] of the single selected Generated instance (CONCEPTS.md
    section 9). The operation's effect is RESOLVED here, at production time: look
    the operation up in the registry by id, evaluate its [set:] expressions with
    the instance's CURRENT params bound under [param], and bake the resulting
    [changes] map into the op (value-in-op). [op_id] also rides on the op as
    journal metadata. [None] unless exactly one Generated element is selected, or
    when the concept/operation is unknown, or when the resolved [changes] map is
    empty. The caller brackets one undo and routes through [Op_apply.op_apply];
    replay merges [changes] and never re-evaluates the expressions. *)
let apply_concept_operation_op (_store : State_store.t) (m : Model.model)
    (op_id : string) : Yojson.Safe.t option =
  let doc = m#document in
  match Document.PathMap.bindings doc.Document.selection with
  | [ (path, _) ] ->
    (match (try Some (Document.get_element doc path) with _ -> None) with
     | Some (Element.Live (Element.Generated gen)) ->
       let concept_id = gen.Element.gen_concept_id in
       let params = gen.Element.gen_params in
       (* Resolve the operation's [set:] expressions over the instance's current
          params, bound under the [param] namespace (the generator's namespace),
          into the concrete [changes] map. Only numeric results are baked, stored
          as floats so serialization matches the conformance corpus. *)
       let changes =
         match Lazy.force workspace with
         | None -> None
         | Some ws ->
           (match Workspace_loader.concept ws concept_id with
            | None -> None
            | Some spec ->
              (match Workspace_loader.json_member "operations" spec with
               | Some (`List ops) ->
                 let operation = List.find_opt (fun o ->
                   match Workspace_loader.json_member "id" o with
                   | Some (`String oid) -> oid = op_id
                   | _ -> false) ops in
                 (match operation with
                  | None -> None
                  | Some operation ->
                    (match Workspace_loader.json_member "set" operation with
                     | Some (`Assoc set_kvs) ->
                       let ctx = `Assoc [ ("param", params) ] in
                       let resolved = List.filter_map (fun (name, expr_v) ->
                         match expr_v with
                         | `String src ->
                           (match Expr_eval.evaluate src ctx with
                            | Expr_eval.Number n -> Some (name, `Float n)
                            | _ -> None)
                         | _ -> None) set_kvs in
                       Some resolved
                     | _ -> None))
               | _ -> None))
       in
       (match changes with
        | Some (_ :: _ as change_kvs) ->
          Some (`Assoc [
            ("op", `String "apply_concept_operation");
            ("path", `List (List.map (fun i -> `Int i) path));
            ("op_id", `String op_id);
            ("changes", `Assoc change_kvs);
          ])
        | _ -> None)
     | _ -> None)
  | _ -> None

let num_of = function
  | `Int i -> float_of_int i
  | `Float f -> f
  | `Intlit s -> float_of_string s
  | _ -> 0.0

(** The render-time concept resolver: given a concept id, return [Some] a
    function that evaluates the concept's generator over a params object and
    yields its [(x, y)] points, or [None] when the concept is unknown. Lets the
    canvas draw a Generated instance's geometry (CONCEPTS.md 3b). *)
let concept_resolver : Live.concept_resolver = fun concept_id ->
  match Lazy.force workspace with
  | None -> None
  | Some ws ->
    (match Workspace_loader.concept ws concept_id with
     | None -> None
     | Some spec ->
       (match Workspace_loader.json_member "generator" spec with
        | Some (`String generator) ->
          Some (fun params ->
            let ctx = `Assoc [ ("param", params) ] in
            match Expr_eval.evaluate generator ctx with
            | Expr_eval.List items ->
              List.filter_map (function
                | `List [ a; b ] -> Some (num_of a, num_of b)
                | _ -> None) items
            | _ -> [])
        | _ -> None))
